src/types/abstract_hmm.jl

```diff
@@ -23,19 +26,29 @@ Any `AbstractHMM` which satisfies the interface can be given to the following fu
 - [`forward_backward`](@ref)
 - [`baum_welch`](@ref) (if `[fit!](@ref)` is implemented)
 """
-abstract type AbstractHMM{ar} end
+abstract type AbstractHMM{T} end
 
 @inline DensityInterface.DensityKind(::AbstractHMM) = HasDensity()
 
-## Interface
 
+############################################################################
+#                              2. INTERFACE                                #
+############################################################################
+
+#------------------------------ 2.1. length -------------------------------#
 """
     length(hmm)
 
 Return the number of states of `hmm`.
 """
 Base.length(hmm::AbstractHMM) = length(initialization(hmm))
+	
+Base.length(hmm::AbstractHMM, control) = length(initialization(hmm, control))
+	
+Base.length(hmm::AbstractHMM, ::Nothing) = length(initialization(hmm))
+
```

Comment:
I know that usually for $n$-dimensional arrays, $n > 1$, it should be `size` and not `length`, but I don't know how to proceed when the initial distribution isn't a $1$-dimensional vector. Canonically for simple MCs/HMMs, it is a $1$-dimensional vector, sure, but for higher-order MCs and input-output HMMs it is a matrix or higher-order "vector" (or, equivalently, a vector of vectors). In the case of higher-order MC, we concluded that the user should tuple-fy in order to make a simple MC with dimension $supp(X_{t})^{l}$... right? But in the case of input-output HMM (or any use of HMM with a control variable), what should be the standard approach?


In the transition matrix and observation distributions we accept the control input, but in the initial distribution it was omitted. This would be immaterial if it wasn't the case that in some functions we do call for the length of the initial distribution (e.g. `viterbi.jl`, line 29).


So, suppose a model in which $\{X_{t}\}$ is a binary parameter process, $\{Y_{t}\}$ is whatever observation process we want, and $\{u_{t}\}$ is a ternary control sequence. There is a initial distribution vector $P(X_{1} \vert u_{1})$ for each one of the $3$ possible values of $u_{1}$, right? Or a rectangular $3 \times 2$ matrix... which is equivalent anyway. If we call `length` in this case, it would return $3$ (or $6$ if we use the matrix), and not $2$, which is the value that we want. This is why I changed the `length` to actually work like `size`, but it was just a initial approach... I totally agree that this isn't a good option, because `length` is expected to work consistently in different objects.


As a $1$-dimensional vector is a particular case of a $n$-dimensional "vector", I thought it would be more abstract if we use `size`. This would enable more diverse cases and, effectively, `size(x, 1)` is the same as `length(x)` for any $1$-dimensional vector, so it wouldn't change anything else. What do you think?